util/singleton.h

```diff
@@ -41,6 +48,55 @@ inline T& ThreadLocalSingleton() {
   thread_local static T instance;
   return instance;
 }
+#else
+template <typename T>
+class ThreadLocalSingletonInternals final {
+ private:
+  static pthread_key_t& KeyStaticStorage() {
+    static pthread_key_t key;
+    return key;
+  }
+
+  static pthread_once_t& KeyIsInitialized() {
+    static pthread_once_t key_is_initialized = PTHREAD_ONCE_INIT;
+    return key_is_initialized;
+  }
+
+  static void Deleter(void* ptr) { delete reinterpret_cast<T*>(ptr); }
+
+  static void CreateKey() {
+    if (pthread_key_create(&(ThreadLocalSingletonInternals::KeyStaticStorage()),
+                           ThreadLocalSingletonInternals::Deleter)) {
+      std::cerr << "Error in `pthread_key_create()`. Terminating." << std::endl;
+      exit(-1);
+    }
+  }
+
+  static pthread_key_t& GetKey() {
+    if (pthread_once(&(ThreadLocalSingletonInternals::KeyIsInitialized()),
+                     ThreadLocalSingletonInternals::CreateKey)) {
+      std::cerr << "Error in `pthread_once()`. Terminating." << std::endl;
+      exit(-1);
+    }
+    return ThreadLocalSingletonInternals::KeyStaticStorage();
+  }
+
+ public:
+  static T& GetInstance() {
+    pthread_key_t& key = ThreadLocalSingletonInternals::GetKey();
+    T* ptr = reinterpret_cast<T*>(pthread_getspecific(key));
+    if (!ptr) {
+      ptr = new T;
```

Comment:
`new T()` would be a tiny bit safer.

(Although the only real usecase I can think of is that with `new T()` a call to `ThreadLocalSingleton<int>` would initialize the `int` with zero, while `new T` leaves it uninitialized.)